src/vs/vscode.d.ts

```diff
@@ -1812,6 +1812,56 @@ declare module 'vscode' {
 	 */
 	export type ProviderResult<T> = T | undefined | null | Thenable<T | undefined | null>;
 
+	/**
+	 * Scope of a code action.
+	 *
+	 * Specifies the type of a code action. A scope is a hierarchical list of identifiers separated by `.`, e.g. `"refactoring.extract.function"`.
+	 */
+	export class CodeActionScope {
+		/**
+		 * Empty scope.
+		 */
+		public static readonly Empty: CodeActionScope;
+
+		/**
+		 * Base scope for refactoring actions.
+		 */
+		public static readonly Refactor: CodeActionScope;
+
+		/**
+		 * Base scope for refactoring extraction actions.
+		 */
+		public static readonly RefactorExtract: CodeActionScope;
+
+		/**
+		 * Base scope for refactoring extraction inline.
+		 */
+		public static readonly RefactorInline: CodeActionScope;
+
+		private constructor(value: string);
+
+		/**
+		 * TODO: should this be exposed?
+		 */
+		readonly value?: string;
+
+		/**
+		 * Create a new scope by appending a more specific selector to the current scope.
+		 *
+		 * Does not modify the current scope object.
+		 */
+		public add(parts: string): CodeActionScope;
+
+		/**
+		 * Does this scope contain scope `other`?
+		 *
+		 * Another way of saying: is `other` within this scope?
+		 *
+		 * @param other Scope to check.
+		 */
+		public contains(other: CodeActionScope): boolean;
```

Comment:
Do we have an use-case for this?

Reply:
Yes, this is used for the `requestedScope` logic in the TS refactoring code action provider so that we can skip computing code actions that vscode will filter out later